### PR TITLE
Forbid identifier lists in measurement assignments

### PR DESCRIPTION
**Pull request by 1ucian0**
After checking with @zachschoenfeld33, it seems like the `quantumMeasurementAssignment` grammar class needs to be adjusted a bit. Currently it looks like this:
```
quantumMeasurementAssignment
    : quantumMeasurement ( ARROW indexIdentifierList)?
    | indexIdentifierList EQUALS quantumMeasurement
    ;
```
In the first line `ARROW indexIdentifierList` should not be optional. In the second line, the idea of multi assignments was discarded and `indexIdentifierList` should be replaced by `indexIdentifier`. The final result is
```
quantumMeasurementAssignment
    : quantumMeasurement ARROW indexIdentifierList
    | indexIdentifier EQUALS quantumMeasurement
    ;
```

This PR adjust the class `quantumMeasurementAssignment` accordingly and adapt the related test.

### CODE REVIEW

File: source/grammar/qasm3.g4 [hunk 1]
```diff
@@ -237,8 +237,8 @@ quantumMeasurement
     ;
 
 quantumMeasurementAssignment
-    : quantumMeasurement ( ARROW indexIdentifierList)?
-    | indexIdentifierList EQUALS quantumMeasurement
+    : quantumMeasurement ARROW indexIdentifierList
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:240]:
```suggestion
    : quantumMeasurement ARROW indexIdentifier
```
Single assignment only

File: source/grammar/qasm3.g4 [hunk 2]
```diff
@@ -237,8 +237,8 @@ quantumMeasurement
     ;
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:237]: The line above `quantumMeasurement`, should only take `indexIdentifier`. That is,
```
quantumMeasurement    
    : 'measure' indexIdentifier
    ;
```

**Comment by zachschoenfeld33** [comment]
@1ucian0 it could be beneficial if you could add a small test to the ANTLR grammar w/ the different measurement types (happy to help w this).

**Comment by 1ucian0** [comment]
> @1ucian0 it could be beneficial if you could add a small test to the ANTLR grammar w/ the different measurement types (happy to help w this).

Sure... I dont fully get the way this is being tested. Maybe in a follow up PR?

File: source/grammar/qasm3.g4 (hunk 2 above)
Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4:237]: @1ucian0 pls update this. We do not allow for measuring multiple individual qubits at once. In order to measure multiple qubits in the same instruction, you'd need to use a qubit register. See https://qiskit.github.io/openqasm/language/insts.html#measurement

File: source/grammar/qasm3.g4 [hunk 3]
```diff
@@ -237,8 +237,8 @@ quantumMeasurement
...
 
 quantumMeasurementAssignment
-    : quantumMeasurement ( ARROW indexIdentifierList)?
-    | indexIdentifierList EQUALS quantumMeasurement
+    : quantumMeasurement ARROW indexIdentifier
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:240]:
```suggestion
    : quantumMeasurement ( ARROW indexIdentifier )?
```
Need this to new optional. Something like `measure a`, without assignment, is valid.

**Comment by zachschoenfeld33** [comment]
> > @1ucian0 it could be beneficial if you could add a small test to the ANTLR grammar w/ the different measurement types (happy to help w this).
> 
> Sure... I dont fully get the way this is being tested. Maybe in a follow up PR?

You can add a line here (https://github.com/Qiskit/openqasm/blob/3cc82855730b512bdab17c3c267a1f4ba3910e6a/source/grammar/tests/outputs/assignment.yaml#L9) that tests the `->` assignment measurement. And then verify that the correct ANTLR is produced.

**Comment by blakejohnson** [comment]
It looks like a few more changes are needed here?

**Comment by taalexander** [comment]
Bump on this @1ucian0?

File: source/grammar/qasm3.g4 (hunk 2 above)
Comment (1ucian0) [reply source/grammar/qasm3.g4:237]: done in 21ee0533ec3ac06bf93d969885ed34ea6e77306b

**Comment by 1ucian0** [comment]
> You can add a line here (
> https://github.com/Qiskit/openqasm/blob/3cc82855730b512bdab17c3c267a1f4ba3910e6a/source/grammar/tests/outputs/assignment.yaml#L9
> 
> ) that tests the `->` assignment measurement. And then verify that the correct ANTLR is produced.

done in 7db94a5a82262757391fc426bc38159c10a78586

**Comment by 1ucian0** [comment]
all change requests addressed.

**Review by taalexander — COMMENTED** [review]
This looks good to me, would you mind taking a look as well @jakelishman?

**Review by jakelishman — APPROVED** [review]
Looks good to me too.  I've just pushed up some tests of invalid measurements to add to this, including ones that are explicitly forbidden by this PR, to catch regressions.

**Pull request merged**